src/Data/FileSystem.elm

```diff
@@ -0,0 +1,186 @@
+            (\ft ->
+                case ft of
+                    File a ->
+                        check a
+
+                    Folder _ _ ->
+                        True
+            )
+
+
+mapFiles : (a -> b) -> FileSystem a -> FileSystem b
+mapFiles f tree =
+    case tree of
+        File x ->
+            File (f x)
+
+        Folder name content ->
+            Folder name (List.map (mapFiles f) content)
+
+
+mapFolders : (List (FileSystem a) -> List (FileSystem a)) -> FileSystem a -> FileSystem a
+mapFolders f tree =
+    case tree of
+        File file ->
+            File file
+
+        Folder name content ->
+            Folder name <| f <| List.map (mapFolders f) content
+
+
+reverse : FileSystem a -> FileSystem a
+reverse =
+    mapFolders List.reverse
+
+
+sortWith : (FileSystem a -> FileSystem a -> Order) -> FileSystem a -> FileSystem a
+sortWith sorter =
+    mapFolders (List.sortWith sorter)
+
+
+sortFoldersAlphabetically : FileSystem a -> FileSystem a
+sortFoldersAlphabetically =
+    sortWith <|
+        \fs1 ->
+            \fs2 ->
+                case ( fs1, fs2 ) of
+                    ( Folder _ _, File _ ) ->
+                        GT
+
+                    ( File _, Folder _ _ ) ->
+                        LT
+
+                    ( Folder name1 _, Folder name2 _ ) ->
+                        compare name1 name2
+
+                    ( File _, File _ ) ->
+                        EQ
+
+
+{-| Interface with BE Audience & AudienceFolder
+
+mkFileSystem transforms BE data into FileSystem of Audiences
+
+-}
+mkFileSystem : List AudienceFolder -> List Audience -> FileSystem Audience
+mkFileSystem folders files =
+    sortAudienceFilesAndFoldersAlphabetically <|
+        mkFileSystemHelper
+            folders
+            (\af -> af.parent == Nothing)
+            files
+            (\a -> a.folder == Nothing)
+            (Folder "ROOT" [])
+
+
+mkFileSystemHelper :
+    List AudienceFolder
+    -> (AudienceFolder -> Bool)
+    -> List Audience
+    -> (Audience -> Bool)
+    -> FileSystem Audience
+    -> FileSystem Audience
+mkFileSystemHelper folders folderFilter files fileFilter tree =
+    case tree of
+        File x ->
+            File x
+
+        Folder name content ->
+            let
+                ( subFolders, restFolders ) =
+                    splitFilter folderFilter folders
+
+                ( subFiles, restFiles ) =
+                    splitFilter fileFilter files
+            in
+            Folder
+                name
+                (content
+                    ++ List.map
+                        (\subFolder ->
+                            mkFileSystemHelper
+                                restFolders
+                                (\af -> af.parent == Just subFolder.id)
+                                restFiles
+                                (\a -> a.folder == Just subFolder.id)
+                                (Folder subFolder.name [])
+                        )
```

Comment:
Q: What do you think about the complexity of this? (no need for a mathematical answer, it can be just practical / gut feeling)

Reply:
I haven't really done any complexity theory yet, but I'd say:
I have to filter through not more than all files and folders for each folder + 1 (the root folder) so
```
let f := length files 
     F := length folders
in  (f+F)*(F+1) ~> F^2 + f*F + F + f
```
So in terms of the of `n := length (files ++ folders)` considering that `n = f + F` hence `f < n` & `F < n` it is better than `n^2 + n^2 + 2n` ~> `2*n^2 + 2*n`

I guess the answer is at worst quadratic and for shallow folder structures much better (close to linear? - for fixed folder structure it is linear `(F+1)*f + (F^2 + F)`).